File: Week 3 & 4/07. Exercises/1. Number Counter/ros2_ws/src/pkg_cpp/src/number_pub.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"

using namespace std::chrono_literals;

class NumberPub : public rclcpp::Node
{
public:
    NumberPub() : Node("number_publisher")
    {
        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number", 10);
        timer_ = this->create_wall_timer(1s, std::bind(&NumberPub::counter, this));
        RCLCPP_INFO(this->get_logger(), "Initial count: %d",count);
    }

private:
    void counter()
    {
        auto msg = example_interfaces::msg::Int64();
        msg.data = count++;
        publisher_->publish(msg);
    }
    int count{};
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char** argv){
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<NumberPub>());
    rclcpp::shutdown();

    return 0;
}

File: Week 5/1. Service/ros2_ws/src/service_act/scripts/number_pub.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64

class NumberPublisher(Node):
    def __init__(self):
        super().__init__("number_publisher")
        self.publisher_ = self.create_publisher(Int64, "number", 10)
        self.timer_ = self.create_timer(0.50, self.counter)
        self.get_logger().info("Counter publisher has started.")
        self.initial_count = 0

    def counter(self, count = Int64()):
        self.initial_count += 1
        count.data = self.initial_count
        # self.get_logger().info(f"Counter: {count.data}")
        self.publisher_.publish(count)


def main(args=None):
    rclpy.init(args=args)
    node = NumberPublisher()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()